### ISSUE TITLE
zenpy hangs

### ISSUE THREAD

**Issue by cblaupanda** [facetoe/zenpy#51]
Hi There 

I have a problem where regularly my zenpy client just hangs without throwing an exception... 

Here my code....

``` Python
ZD_EXCEPTIONS = (
    ZenpyException,
    APIException,
    RecordNotFoundException,
    ZenpyCacheException
    )

try:
    usr = []
    for obj in zenpy.group_memberships(timeout=0.05):
        if obj.user.role == 'agent' and \
          obj.user.to_dict()['user_fields']['custom_name'] is not None:
            data = [
                obj.user.to_dict()['user_fields']['custom_name'].replace(
                    '.domain.com', '').replace('-', '_'),
                obj.user.id,
                obj.user.email,
                obj.user.organization.id,
                obj.user.organization.name,
                ]
            if data not in usr:
                usr.append(data)
except ZD_EXCEPTIONS as error:
    print('Problem connecting to Zendesk API. {}'.format(error))
    sys.exit(1)
```

I assume I am hitting rate limit or something but it is not caught by the timeout either. I found this article regarding best practices on rate limit. https://support.zendesk.com/hc/en-us/articles/203691336-Best-practices-for-avoiding-rate-limiting

Any suggestions what I can do?

Many Thanks
Christoph

**Comment by facetoe** [comment]
Hi @cblaupanda, 

If you are hitting the rate limit Zenpy will wait the required period of time before [retrying](https://github.com/facetoe/zenpy/blob/master/zenpy/lib/api.py#L54) the request and it will also log that this is happening. As a first step it would be a good idea to enable logging to verify that you are being rate limited. Try adding the following to your code above and see if it reports being rate limited:

```python
log = logging.getLogger()
log.setLevel(logging.DEBUG)
ch = logging.StreamHandler(sys.stdout)
formatter = logging.Formatter('%(levelname)s - %(message)s')
ch.setFormatter(formatter)
log.addHandler(ch)
logging.getLogger("requests").setLevel(logging.WARNING)
```
The `timeout` keyword that you are passing to the `group_memberships` endpoint is ignored and has no effect.

**Comment by cblaupanda** [comment]
it hang now for more than 30 minutes so I hit Ctrl-C. Here the output:
```
^CTraceback (most recent call last):
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 376, in _make_request
    httplib_response = conn.getresponse(buffering=True)
TypeError: getresponse() got an unexpected keyword argument 'buffering'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./zendesk_user_sync.py", line 226, in <module>
    if obj.user.role == 'agent' and \
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/zenpy/lib/api_objects.py", line 200, in user
    return self.api.get_user(self.user_id)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/zenpy/lib/api.py", line 182, in get_user
    return self._get_item(_id, endpoint, object_type, sideload=True)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/zenpy/lib/api.py", line 98, in _get_item
    _json = self._query(endpoint=endpoint(id=_id, sideload=sideload))
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/zenpy/lib/api.py", line 114, in _query
    response = self._get(self._get_url(endpoint=endpoint))
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/zenpy/lib/api.py", line 56, in _get
    response = self.session.get(url, stream=stream)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/sessions.py", line 480, in get
    return self.request('GET', url, **kwargs)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/sessions.py", line 468, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/adapters.py", line 376, in send
    timeout=timeout
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 559, in urlopen
    body=body, headers=headers)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 378, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.4/http/client.py", line 1171, in getresponse
    response.begin()
  File "/usr/lib/python3.4/http/client.py", line 351, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.4/http/client.py", line 313, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.4/socket.py", line 374, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3.4/ssl.py", line 769, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/lib/python3.4/ssl.py", line 641, in read
    v = self._sslobj.read(len, buffer)
KeyboardInterrupt
```

I will test with the logging enabled and let you know.

Many thanks!

**Comment by facetoe** [comment]
Hmm looks like it is hanging in the `_get` method that handles rate limiting. Let me know what you find when logging is enabled, there could be a bug in how Zenpy handles the rate limiting as it's not something I've hit often.

**Comment by cblaupanda** [comment]
it just hangs here...

```
DEBUG - Cache MISS: [User 2562947935]
DEBUG - GET: https://company.zendesk.com/api/v2/users/2562947935.json&include=organizations,abilities,roles,identities,groups
DEBUG - Caching: [User 2562947935]
DEBUG - Cache HIT: [User 2562947935]
DEBUG - Cache HIT: [User 2562947935]
DEBUG - Cache MISS: [User 2562947945]
DEBUG - GET: https://company.zendesk.com/api/v2/users/2562947945.json&include=organizations,abilities,roles,identities,groups
```

**Comment by facetoe** [comment]
Hmm, doesn't look like you're being rate limited. If you paste that URL in a browser does it also hang?

**Comment by cblaupanda** [comment]
okay.... here I just got the exception without cancelling the script. 

```Python
DEBUG - Cache MISS: [User 1384981551]
DEBUG - GET: https://company.zendesk.com/api/v2/users/1384981551.json&include=organizations,abilities,roles,identities,groups
Traceback (most recent call last):
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 376, in _make_request
    httplib_response = conn.getresponse(buffering=True)
TypeError: getresponse() got an unexpected keyword argument 'buffering'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 559, in urlopen
    body=body, headers=headers)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 378, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.4/http/client.py", line 1171, in getresponse
    response.begin()
  File "/usr/lib/python3.4/http/client.py", line 351, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.4/http/client.py", line 321, in _read_status
    raise BadStatusLine(line)
http.client.BadStatusLine: ''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/adapters.py", line 376, in send
    timeout=timeout
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 609, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/packages/urllib3/util/retry.py", line 247, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/packages/urllib3/packages/six.py", line 309, in reraise
    raise value.with_traceback(tb)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 559, in urlopen
    body=body, headers=headers)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 378, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.4/http/client.py", line 1171, in getresponse
    response.begin()
  File "/usr/lib/python3.4/http/client.py", line 351, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.4/http/client.py", line 321, in _read_status
    raise BadStatusLine(line)
requests.packages.urllib3.exceptions.ProtocolError: ('Connection aborted.', BadStatusLine("''",))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./zendesk_user_sync.py", line 244, in <module>
    if obj.user.role == 'agent' and \
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/zenpy/lib/api_objects.py", line 200, in user
    return self.api.get_user(self.user_id)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/zenpy/lib/api.py", line 182, in get_user
    return self._get_item(_id, endpoint, object_type, sideload=True)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/zenpy/lib/api.py", line 98, in _get_item
    _json = self._query(endpoint=endpoint(id=_id, sideload=sideload))
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/zenpy/lib/api.py", line 114, in _query
    response = self._get(self._get_url(endpoint=endpoint))
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/zenpy/lib/api.py", line 56, in _get
    response = self.session.get(url, stream=stream)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/sessions.py", line 480, in get
    return self.request('GET', url, **kwargs)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/sessions.py", line 468, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/share/python/zendesk-user-sync/lib/python3.4/site-packages/requests/adapters.py", line 426, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', BadStatusLine("''",))

```

**Comment by facetoe** [comment]
I think this is an error in the requests library not Zenpy. I found a similar bug report here - https://github.com/kennethreitz/requests/issues/2364. Are you behind a proxy at all?

**Comment by kyledemeule** [comment]
I'm not sure if this is the same issue, but I'm seeing the same exception raised after making a couple hundred calls. E.g.
```
    job_status = zen.organizations.create(to_upload)
  File "build/bdist.linux-x86_64/egg/zenpy/lib/api.py", line 309, in create
  File "build/bdist.linux-x86_64/egg/zenpy/lib/api.py", line 292, in _do
  File "build/bdist.linux-x86_64/egg/zenpy/lib/api.py", line 43, in post
  File "/home/vagrant/.local/lib/python2.7/site-packages/requests/sessions.py", line 522, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/home/vagrant/.local/lib/python2.7/site-packages/requests/sessions.py", line 475, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/vagrant/.local/lib/python2.7/site-packages/requests/sessions.py", line 596, in send
    r = adapter.send(request, **kwargs)
  File "/home/vagrant/.local/lib/python2.7/site-packages/requests/adapters.py", line 473, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', BadStatusLine("''",))
```
and
```
    job_result = zen.job_status(id=job_status.id)
  File "build/bdist.linux-x86_64/egg/zenpy/lib/api.py", line 174, in __call__
  File "build/bdist.linux-x86_64/egg/zenpy/lib/api.py", line 77, in _get_items
  File "build/bdist.linux-x86_64/egg/zenpy/lib/api.py", line 100, in _get_item
  File "build/bdist.linux-x86_64/egg/zenpy/lib/api.py", line 116, in _query
  File "build/bdist.linux-x86_64/egg/zenpy/lib/api.py", line 54, in _get
  File "build/bdist.linux-x86_64/egg/zenpy/lib/api.py", line 58, in _call_api
  File "/home/vagrant/.local/lib/python2.7/site-packages/requests/sessions.py", line 488, in get
    return self.request('GET', url, **kwargs)
  File "/home/vagrant/.local/lib/python2.7/site-packages/requests/sessions.py", line 475, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/vagrant/.local/lib/python2.7/site-packages/requests/sessions.py", line 596, in send
    r = adapter.send(request, **kwargs)
  File "/home/vagrant/.local/lib/python2.7/site-packages/requests/adapters.py", line 473, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', BadStatusLine("''",))
```

@facetoe do you have any suggestion? I'm not behind a proxy, and I tried this on two different internet connections (office and home).

**Comment by facetoe** [comment]
Hi @kyledemeule, I think this is the same bug as reported here - https://github.com/kennethreitz/requests/issues/2364. In https://github.com/kennethreitz/requests/issues/3661 a workaround is suggested that will cause requests to transparently retry the connections under the hood when this condition arises. Are you able to test  a6ee35d and let me know if it works?

**Comment by kyledemeule** [comment]
Hi @facetoe , thanks for the quick reply. Unfortunately I had a large job I was doing, and already completed it before seeing your solution. The workaround I used was something like this:
```
def wrapped_zenpy_call(zenpy_call, *args, **kwargs):
    count = 0
    while count < 10:
        try:
            return zenpy_call(*args, **kwargs)
        except requests.exceptions.ConnectionError:
            print "Connection error {}.".format(count)
            if count < 10:
                count += 1
            else:
                raise
```
I never had a call fail more than once in a row, so anything using retries would probably work. Unfortunately I can't run my job again and test your solution.

Thanks!

**Comment by facetoe** [comment]
No worries, thanks for the info!

**Comment by facetoe** [comment]
I ended up merging the fix in the hope it fixes this issue (#82)
